src/PowerShellEditorServices/Services/DebugAdapter/BreakpointSyncService.cs

```diff
@@ -0,0 +1,904 @@
+
+    private readonly ILanguageServerFacade _languageServer;
+
+    private bool? _isSupported;
+
+    public BreakpointSyncService(
+        IInternalPowerShellExecutionService host,
+        ILoggerFactory loggerFactory,
+        DebugStateService debugStateService,
+        ILanguageServerFacade languageServer)
+    {
+        _host = host;
+        _logger = loggerFactory.CreateLogger<BreakpointSyncService>();
+        _debugStateService = debugStateService;
+        _languageServer = languageServer;
+    }
+
+    public bool IsSupported
+    {
+        get => _isSupported ?? false;
+        set
+        {
+            Debug.Assert(_isSupported is null, "BreakpointSyncService.IsSupported should only be set once by initialize.");
+            _isSupported = value;
+        }
+    }
+
+    public bool IsMutatingBreakpoints
+    {
+        get
+        {
+            if (_breakpointMutateHandle.Wait(0))
+            {
+                _breakpointMutateHandle.Release();
+                return false;
+            }
+
+            return true;
+        }
+    }
+
+    private int? TargetRunspaceId => _debugStateService?.RunspaceId;
+
+    public async Task UpdatedByServerAsync(BreakpointUpdatedEventArgs eventArgs)
+    {
+        if (!_map.ByPwshId.TryGetValue(eventArgs.Breakpoint.Id, out SyncedBreakpoint existing))
+        {
+            // If we haven't told the client about the breakpoint yet then we can just ignore.
+            if (eventArgs.UpdateType is BreakpointUpdateType.Removed)
+            {
+                return;
+            }
+
+            existing = CreateFromServerBreakpoint(eventArgs.Breakpoint);
+            string? id = await SendToClientAsync(existing.Client, BreakpointUpdateType.Set)
+                .ConfigureAwait(false);
+
+            existing.Client.Id = id!;
+            RegisterBreakpoint(existing);
+            return;
+        }
+
+        if (eventArgs.UpdateType is BreakpointUpdateType.Removed)
+        {
+            UnregisterBreakpoint(existing);
+            await SendToClientAsync(existing.Client, BreakpointUpdateType.Removed).ConfigureAwait(false);
+            return;
+        }
+
+        if (eventArgs.UpdateType is BreakpointUpdateType.Enabled or BreakpointUpdateType.Disabled)
+        {
+            await SendToClientAsync(existing.Client, eventArgs.UpdateType).ConfigureAwait(false);
+            bool isActive = eventArgs.UpdateType is BreakpointUpdateType.Enabled;
+            SyncedBreakpoint newBreakpoint = existing with
+            {
+                Client = existing.Client with
+                {
+                    Enabled = isActive,
+                },
+            };
+
+            UnregisterBreakpoint(existing);
+            RegisterBreakpoint(newBreakpoint);
+            return;
+        }
+
+        LogBreakpointError(
+            existing,
+            "Somehow we're syncing a new breakpoint that we've already sync'd. That's not supposed to happen.");
```

Comment:
You _know_ we're gonna see this in a user's log one day and go 😳

Reply:
> You _know_ we're gonna see this in a user's log one day and go 😳

oh yeah 💯 

But we're gonna have a whole lot of info when it happens! (that method writes out a bunch of info about the bp)

Reply:
I saw! It's wonderful.